dChatServer/ChatPacketHandler.cpp

```diff
@@ -398,72 +386,83 @@ void ChatPacketHandler::HandlePrivateChatMessage(Packet* packet) {
+	}
 
-	//To the sender:
-	{
-		CBITSTREAM;
-		BitStreamUtils::WriteHeader(bitStream, eConnectionType::CHAT_INTERNAL, eChatInternalMessageType::ROUTE_TO_PLAYER);
-		bitStream.Write(sender.playerID);
+	if (sender.isFTP) {
+		SendPrivateChatMessage(sender, receiver, sender, message, eChatMessageResponseCode::SENDERFREETRIAL);
+		return;
+	}
 
-		BitStreamUtils::WriteHeader(bitStream, eConnectionType::CHAT, eChatMessageType::PRIVATE_CHAT_MESSAGE);
-		bitStream.Write(sender.playerID);
-		bitStream.Write(eChatChannel::PRIVATE_CHAT);
-		bitStream.Write<uint32_t>(0); // not used
-		bitStream.Write(LUWString(sender.playerName));
-		bitStream.Write(sender.playerID);
-		bitStream.Write<uint16_t>(0); // sourceID
-		bitStream.Write(sender.GMLevel);
-		bitStream.Write(LUWString(receiver.playerName));
-		bitStream.Write(receiver.GMLevel);
-		bitStream.Write(eChatMessageResponseCode::Sent);
-		bitStream.Write(message);
+	if (receiver.isFTP) {
+		SendPrivateChatMessage(sender, receiver, sender, message, eChatMessageResponseCode::RECEIVERFREETRIAL);
+		return;
+	}
 
-		SystemAddress sysAddr = sender.sysAddr;
-		SEND_PACKET;
+	// Check to see if they are friends
+	// only freinds can whispr each other
+	for (const auto& fr : receiver.friends) {
+		if (fr.friendID == sender.playerID) {
+			//To the sender:
+			SendPrivateChatMessage(sender, receiver, sender, message, eChatMessageResponseCode::SENT);
+			//To the receiver:
+			SendPrivateChatMessage(sender, receiver, receiver, message, eChatMessageResponseCode::RECEIVEDNEWWHISPER);
+			return;
+		}
 	}
+	SendPrivateChatMessage(sender, receiver, sender, message, eChatMessageResponseCode::NOTFRIENDS);
+}
 
-	//To the receiver:
-	{
-		CBITSTREAM;
-		BitStreamUtils::WriteHeader(bitStream, eConnectionType::CHAT_INTERNAL, eChatInternalMessageType::ROUTE_TO_PLAYER);
-		bitStream.Write(receiver.playerID);
+void ChatPacketHandler::SendPrivateChatMessage(const PlayerData& sender, const PlayerData& receiver, const PlayerData& routeTo, const LUWString& message, eChatMessageResponseCode responseCode) {
+	CBITSTREAM;
+	BitStreamUtils::WriteHeader(bitStream, eConnectionType::CHAT_INTERNAL, eChatInternalMessageType::ROUTE_TO_PLAYER);
+	bitStream.Write(routeTo.playerID);
 
-		BitStreamUtils::WriteHeader(bitStream, eConnectionType::CHAT, eChatMessageType::PRIVATE_CHAT_MESSAGE);
-		bitStream.Write(sender.playerID);
-		bitStream.Write(eChatChannel::PRIVATE_CHAT);
-		bitStream.Write<uint32_t>(0); // not used
-		bitStream.Write(LUWString(sender.playerName));
-		bitStream.Write(sender.playerID);
-		bitStream.Write<uint16_t>(0); // sourceID
-		bitStream.Write(sender.GMLevel);
-		bitStream.Write(LUWString(receiver.playerName));
-		bitStream.Write(receiver.GMLevel);
-		bitStream.Write(eChatMessageResponseCode::ReceivedNewWhisper);
-		bitStream.Write(message);
+	BitStreamUtils::WriteHeader(bitStream, eConnectionType::CHAT, eChatMessageType::PRIVATE_CHAT_MESSAGE);
+	bitStream.Write(sender.playerID);
+	bitStream.Write(eChatChannel::TEAM);
```

Comment:
This function is called SendPrivateChatmessage, but the chat channel is Team.  Why was this changed from what it was previously?

Reply:
yep, it should take in the channel to send on as well. oversight on my part
teams is technically a type of private chat message as well